Tests/BSONTests/DocumentTests.swift

```diff
@@ -892,7 +715,8 @@ final class DocumentTests: MongoSwiftTestCase {
             try container.encode(customEncodedDoc)
         }
         decoder.dataDecodingStrategy = .custom { decoder in
-            let doc = try Document(from: decoder)
+            let container = try decoder.singleValueContainer()
```

Comment:
@patrickfreed, this is related to our conversation on a recent PR: https://github.com/mongodb/swift-bson/pull/19#discussion_r451186379

since this code was directly calling through to `Document`'s `Codable` conformance (even though it was only ever happening with a `BSONDecoder`), we would end up calling the default `BSONValue.init(from decoder: Decoder)` implementation (which just calls `getDecodingError`, since we assumed no one would ever hit it while using a `BSONDecoder`) and then getting a decoding error. 

very confusingly, the error here was a type mismatch where both the expected and actual type were the same (`BSONDocument`), because we would hit this logic: https://github.com/mongodb/swift-bson/blob/aa44622750207f305dbd16769adae552255ece65/Sources/BSON/BSONError.swift#L140

this makes me think that maybe we need to revisit the idea of just erroring in that method and maybe need to check for decoder type, because it's possible even when using a decoder that we own to hit such an error.

I've updated to this workaround for now which goes through the decoder's container API and therefore hits our special-casing.

though, thinking about it more, I actually think it's unwise to use these initializers in general, because they can inadvertently bypass decoding strategies where the default implementation of the method is not used at all. 
for example, the way we implement our `Date` decoding strategies (and the way I believe `JSONDecoder` does it too) relies on seeing a call to decode a `Date` from a decoding container. We then switch according to the strategy, and only actually use that method at all in the `.deferredToDate` case:
https://github.com/mongodb/swift-bson/blob/aa44622750207f305dbd16769adae552255ece65/Sources/BSON/BSONDecoder.swift#L400

while the `encode(to: encoder)` and `init(from: decoder)` methods are public, I think they're actually not really intended for consumption except for within encoders and decoders.

Reply:
Nice catch, and I think you're right about avoiding direct usage of the `Codable` requirements, so I don't think what you've done is really a workaround but in fact is just the right way to do it.

Reply:
that said, although it is unwise to do this, it's an easy mistake to make, and the error we currently output is incorrect / unhelpful. 

I'm wondering if we should handle this in the default conformances on BSONValue like:
```swift
    public init(from decoder: Decoder) throws {
        guard let bsonDecoder = decoder as? _BSONDecoder else {
            throw getDecodingError(type: Self.self, decoder: decoder)
        }

        let topContainer = bsonDecoder.storage.topContainer
        guard let value = topContainer.bsonValue as? Self else {
            throw DecodingError._typeMismatch(at: decoder.codingPath, expectation: Self.self, reality: topContainer.bsonValue)
        }
        self = value
    }
```

```swift
    public func encode(to encoder: Encoder) throws {
        guard let bsonEncoder = encoder as? _BSONEncoder else {
            throw bsonEncodingUnsupportedError(value: self, at: encoder.codingPath)
        }
        bsonEncoder.storage.containers.append(self)
    }
```

what do you think?

alternatively we could just throw a different error message if it's a BSON/extJSON encoder/decoder vs any other kind.

Reply:
Yeah this seems like a good approach.

Reply:
Is this something we should do here, or as part of follow on work?

Reply:
I opened [SWIFT-967](https://jira.mongodb.org/browse/SWIFT-967).